pkg/disk/disk.go

```diff
@@ -0,0 +1,254 @@
+		InstanceManagerDiskServiceAPIMinVersion: int64(v.InstanceManagerDiskServiceAPIMinVersion),
+	}, nil
+}
+
+func (s *Server) DiskCreate(ctx context.Context, req *rpc.DiskCreateRequest) (*rpc.Disk, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType":  req.DiskType,
+		"diskName":  req.DiskName,
+		"diskPath":  req.DiskPath,
+		"blockSize": req.BlockSize,
+	})
+
+	log.Info("Creating disk")
+
+	if req.DiskName == "" || req.DiskPath == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk path are required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		ret, err := s.spdkClient.DiskCreate(req.DiskName, req.DiskPath, req.BlockSize)
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+		}
+		return spdkDiskToDisk(ret), nil
+	case rpc.DiskType_filesystem:
+		// TODO: implement filesystem disk type
+		fallthrough
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.Unimplemented, "unsupported disk type %v", req.DiskType)
+	}
+}
+
+func (s *Server) DiskDelete(ctx context.Context, req *rpc.DiskDeleteRequest) (*emptypb.Empty, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskName": req.DiskName,
+		"diskUUID": req.DiskUuid,
+	})
+
+	log.Info("Deleting disk")
+
+	if req.DiskName == "" || req.DiskUuid == "" {
+		return &empty.Empty{}, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk UUID are required")
+	}
+
+	if s.spdkClient != nil {
+		return &empty.Empty{}, s.spdkClient.DiskDelete(req.DiskName, req.DiskUuid)
+	}
+
+	return &empty.Empty{}, nil
+}
+
+func (s *Server) DiskGet(ctx context.Context, req *rpc.DiskGetRequest) (*rpc.Disk, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType": req.DiskType,
+		"diskName": req.DiskName,
+		"diskPath": req.DiskPath,
+	})
+
+	log.Info("Getting disk info")
+
+	if req.DiskName == "" || req.DiskPath == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk path are required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		ret, err := s.spdkClient.DiskGet(req.DiskName)
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+		}
+		return spdkDiskToDisk(ret), nil
+	case rpc.DiskType_filesystem:
+		// TODO: implement filesystem disk type
+		fallthrough
+	default:
+		return nil, grpcstatus.Errorf(grpccodes.Unimplemented, "unsupported disk type %v", req.DiskType)
+	}
+}
+
+func (s *Server) DiskReplicaInstanceList(ctx context.Context, req *rpc.DiskReplicaInstanceListRequest) (*rpc.DiskReplicaInstanceListResponse, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType": req.DiskType,
+		"diskName": req.DiskName,
+	})
+
+	log.Info("Listing disk replica instances")
+
+	if req.DiskName == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name is required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		replicas, err := s.spdkClient.ReplicaList()
```

Comment:
Shouldn't the disk name be used here? the code here is just listing all replicas from the spdk engine.

Reply:
We have a replicaMap (cache) where replica names serve as keys, so the disk name is not used.

Reply:
We have a replicaMap (cache) where replica names server as keys.
ReplicaList doesn't support filter by diskName yet.